parla/parray/core.py

```diff
@@ -14,6 +16,241 @@
     cupy = numpy
     gpu = None
 
+
+CPU_INDEX = -1
+
+
+class MemoryOperation:
+    """
+    A memory operation representation.
+    """
+    ERROR = -1  # there is an error
+    NOOP = 0    # no operation
+    LOAD = 1    # load data from src to dst
+    EVICT = 2   # clear the data in src
+
+    def __init__(self, operation: int = NOOP, dst: int = -1, src: int = -1):
+        self.operation = operation
+        self.dst = dst
+        self.src = src
+
+    @staticmethod
+    def noop() -> MemoryOperation:
+        return MemoryOperation()
+
+    @staticmethod
+    def error() -> MemoryOperation:
+        return MemoryOperation(MemoryOperation.ERROR)
+
+    @staticmethod
+    def load(dst, src) -> MemoryOperation:
+        return MemoryOperation(MemoryOperation.LOAD, dst, src)
+
+    @staticmethod
+    def evict(src) -> MemoryOperation:
+        return MemoryOperation(MemoryOperation.EVICT, src=src)
+
+
+class Coherence:
+    """
+    A memory coherence protocol between devices.
+
+    Implements MSI protocol.
+    """
+    INVALID = 0
+    SHARED = 1
+    MODIFIED = 2
+
+    def __init__(self, init_owner: int):
+        """
+        Args:
+            init_owner: the owner of the first copy in the system
+        """
+        self._coherence_states = {init_owner: self.SHARED}  # states of each device
+        self._owner = init_owner  # owner id when MODIFIED / smallest valid device id when SHARED
+        self._overall_state = self.SHARED  # state of the whole system
+
+    def register_device(self, operator):
+        """ Register a new device to the protocol.
+
+        Should be called before do other operations as an operator.
+
+        Args:
+            operator: device id of the new device
+        """
+        self._coherence_states[operator] = self.INVALID
+
+    def update(self, operator, do_write=False) -> List[MemoryOperation]:
+        """ Tell the protocol that operator get a new copy (e.g. from user).
+
+        Should be called before do other operations as an operator.
+
+        Args:
+            operator: device id of the new device
+            do_write: if True, the operator will be MODIFIED, otherwise SHARED
+        """
+        operator_state = self._coherence_states[operator]
+
+        if operator_state == self.MODIFIED:  # already have the right to write data
+            return [MemoryOperation.noop()]
+        else:
+            if self._overall_state == self.INVALID:  # the system doesn't hold a copy before
+                self._coherence_states[operator] = self.MODIFIED if do_write else self.SHARED
+                self._owner = operator
+                self._overall_state = self.MODIFIED if do_write else self.SHARED
+                return [MemoryOperation.noop()]
+            else:  # the system already hold a copy
+                # evict others
```

Comment:
Just a small comment: the evict others pattern is used a lot, maybe it can be refactored into a single function (use evict from below)?

Reply:
The reason I make it a separate one is because the detail is slightly different (the if statement). The Coherence.evict() is not used currently. It is reserved for future use (for example, scheduler could free up GPU memory by call evict on some data)